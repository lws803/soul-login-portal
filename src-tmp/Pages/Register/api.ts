import axios from 'axios';

import { BASE_URL } from '../../constants';
import { ApiError } from '../../shared/apiTypes';

export const register = async ({
  values,
}: {
  values: { email: string; password: string; username: string };
}) => {
  try {
    const { data } = await axios.post<Response>(`${BASE_URL}/v1/users`, values);
    return { data, error: null };
  } catch ({ response: { data, status } }) {
    return {
      data: null,
      error: { data: data as ApiError, status: status as number },
    };
  }
};

type Response = {
  id?: number;
};
