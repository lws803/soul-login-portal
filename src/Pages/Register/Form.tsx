import { useState } from 'react';
import {
  FormControl,
  FormLabel,
  Input,
  FormHelperText,
  FormErrorMessage,
  Box,
  Progress,
} from '@chakra-ui/react';
import * as Yup from 'yup';
import zxcvbn from 'zxcvbn';
import { useFormik } from 'formik';
import { useNavigate, useLocation } from 'react-router-dom';

import FancyButton from 'src/components/FancyButton';

import { register } from './api';

export default function Form({ setErrors }: Props) {
  const navigate = useNavigate();
  const { search } = useLocation();

  const [isRegistering, setIsRegistering] = useState(false);

  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
      username: '',
    },
    onSubmit: async (values) => {
      setErrors([]);
      setIsRegistering(true);

      const { data, error } = await register({ values });
      if (error) {
        setIsRegistering(false);
        if (error.status === 400 && error.data.error === 'VALIDATION_ERROR') {
          formik.setErrors({
            password: 'Password is too weak!',
          });
        } else if (error.status === 409) {
          formik.setErrors({ email: error.data.message });
        } else {
          setErrors([error.data.message]);
        }
      }
      if (data?.id) {
        navigate(`/${search}`);
      }
    },
    validationSchema: Yup.object({
      email: Yup.string().email().required(),
      username: Yup.string().required(),
      password: Yup.string().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.username && formik.touched.username}
      >
        <FormLabel htmlFor="username">Username</FormLabel>
        <Input
          id="username"
          name="username"
          onChange={formik.handleChange}
          value={formik.values.username}
          aria-label="Username input"
          variant="filled"
          placeholder="Your username"
          disabled={isRegistering}
        />
        {!formik.errors.email && (
          <FormHelperText>Choose an awesome username!</FormHelperText>
        )}
        {formik.errors.username && (
          <FormErrorMessage>{formik.errors.username}</FormErrorMessage>
        )}
      </FormControl>

      <FormControl
        isInvalid={!!formik.errors.email && formik.touched.email}
        marginTop={8}
      >
        <FormLabel htmlFor="email">Email</FormLabel>
        <Input
          id="email"
          name="email"
          onChange={formik.handleChange}
          value={formik.values.email}
          aria-label="Email input"
          variant="filled"
          placeholder="Your email"
          disabled={isRegistering}
        />
        {!formik.errors.email && (
          <FormHelperText>We&apos;ll never share your email.</FormHelperText>
        )}
        {formik.errors.email && (
          <FormErrorMessage>{formik.errors.email}</FormErrorMessage>
        )}
      </FormControl>
      <FormControl
        isInvalid={!!formik.errors.password && formik.touched.password}
        marginTop={8}
      >
        <FormLabel htmlFor="password">Password</FormLabel>
        {formik.values.password.length > 0 && (
          <Progress
            value={((zxcvbn(formik.values.password).score + 1) / 5) * 100}
            colorScheme="soul.green"
            size="xs"
            mb={1}
          />
        )}
        <Input
          id="password"
          name="password"
          type="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          aria-label="Password input"
          variant="filled"
          placeholder="Your password"
          disabled={isRegistering}
        />
        {formik.errors.password && (
          <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
        )}
      </FormControl>
      <Box mt={8}>
        <FancyButton
          isLoading={isRegistering}
          disabled={isRegistering}
          type="submit"
        >
          Register
        </FancyButton>
      </Box>
    </form>
  );
}

type Props = {
  setErrors: (errors: string[]) => void;
};
