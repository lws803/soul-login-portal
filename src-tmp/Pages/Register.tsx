import { useState } from 'react';
import { Text, Link } from '@chakra-ui/react';
import { useLocation } from 'react-router-dom';

import Form from './Register/Form';

import Page from '../components/Page';
import Title from '../components/Title';

export default function Register() {
  const [errors, setErrors] = useState<string[]>([]);
  const { search } = useLocation();

  return (
    <Page errors={errors}>
      <Title
        title="Create new account"
        subTitle={
          <Text>
            Already a member?{' '}
            <Link href={`/${search}`} target="_self" whiteSpace="nowrap">
              Log in
            </Link>
          </Text>
        }
      />
      <Form setErrors={setErrors} />
    </Page>
  );
}
