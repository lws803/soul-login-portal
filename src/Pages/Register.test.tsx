import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import Register from './Register';

describe(Register, () => {
  const path = '/?platformId=1&callback=https://www.example.com';

  it('renders', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[path]}>
        <Register />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });
});
