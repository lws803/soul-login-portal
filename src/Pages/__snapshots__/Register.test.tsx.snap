// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1wiwcvy"
    >
      <div
        class="css-1ryq4o3"
      >
        <div
          class="css-1yjvs5a"
        >
          <p
            class="chakra-text css-odt1ir"
          >
            Create new account
          </p>
          <div
            class="css-164r41r"
          >
            <p
              class="chakra-text css-0"
            >
              Already a member?
               
              <a
                class="chakra-link css-epvm6"
                href="/?platformId=1&callback=https://www.example.com&state=state&codeChallenge=CODE_CHALLENGE"
                target="_self"
              >
                <strong>
                  Log in
                </strong>
              </a>
            </p>
          </div>
        </div>
        <form>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="username"
              id="field-:r0:-label"
            >
              Username
            </label>
            <input
              aria-describedby="field-:r0:-helptext"
              aria-label="Username input"
              class="chakra-input css-0"
              id="username"
              name="username"
              placeholder="Your username"
              value=""
            />
            <div
              class="chakra-form__helper-text css-0"
              id="field-:r0:-helptext"
            >
              Choose an awesome username!
            </div>
          </div>
          <div
            class="chakra-form-control css-164r41r"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="email"
              id="field-:r1:-label"
            >
              Email
            </label>
            <input
              aria-describedby="field-:r1:-helptext"
              aria-label="Email input"
              class="chakra-input css-0"
              id="email"
              name="email"
              placeholder="Your email"
              value=""
            />
            <div
              class="chakra-form__helper-text css-0"
              id="field-:r1:-helptext"
            >
              We'll never share your email.
            </div>
          </div>
          <div
            class="chakra-form-control css-164r41r"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="password"
              id="field-:r2:-label"
            >
              Password
            </label>
            <input
              aria-label="Password input"
              class="chakra-input css-0"
              id="password"
              name="password"
              placeholder="Your password"
              type="password"
              value=""
            />
          </div>
          <div
            class="css-164r41r"
          >
            <button
              class="css-y6mgpv"
              type="submit"
            >
              <div
                class="css-tkevr6"
              >
                Register
              </div>
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="css-l5skqa"
    >
      <div
        class="css-18hh56p"
      >
        <div
          class="chakra-stack css-1o71ot0"
        >
          <a
            class="chakra-link css-epvm6"
            href="https://api.soul-network.com/docs"
            target="_blank"
          >
            API Docs
          </a>
          <a
            class="chakra-link css-epvm6"
            href="https://github.com/soul-project/login-portal"
            target="_blank"
          >
            Github
          </a>
          <a
            class="chakra-link css-epvm6"
            href="https://github.com/soul-project/soul"
            target="_blank"
          >
            Soul Network's Github
          </a>
          <a
            class="chakra-link css-epvm6"
            href="mailto:lws803@gmail.com"
            target="_blank"
          >
            Contact me
          </a>
        </div>
        <p
          class="chakra-text css-1m360k3"
        >
          Copyright © 
          2022
           Soul Network.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Register renders without path 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1wiwcvy"
    >
      <div
        class="css-1ryq4o3"
      >
        <div
          class="css-1yjvs5a"
        >
          <p
            class="chakra-text css-odt1ir"
          >
            Create new account
          </p>
          <div
            class="css-164r41r"
          />
        </div>
        <form>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="username"
              id="field-:r3:-label"
            >
              Username
            </label>
            <input
              aria-describedby="field-:r3:-helptext"
              aria-label="Username input"
              class="chakra-input css-0"
              id="username"
              name="username"
              placeholder="Your username"
              value=""
            />
            <div
              class="chakra-form__helper-text css-0"
              id="field-:r3:-helptext"
            >
              Choose an awesome username!
            </div>
          </div>
          <div
            class="chakra-form-control css-164r41r"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="email"
              id="field-:r4:-label"
            >
              Email
            </label>
            <input
              aria-describedby="field-:r4:-helptext"
              aria-label="Email input"
              class="chakra-input css-0"
              id="email"
              name="email"
              placeholder="Your email"
              value=""
            />
            <div
              class="chakra-form__helper-text css-0"
              id="field-:r4:-helptext"
            >
              We'll never share your email.
            </div>
          </div>
          <div
            class="chakra-form-control css-164r41r"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="password"
              id="field-:r5:-label"
            >
              Password
            </label>
            <input
              aria-label="Password input"
              class="chakra-input css-0"
              id="password"
              name="password"
              placeholder="Your password"
              type="password"
              value=""
            />
          </div>
          <div
            class="css-164r41r"
          >
            <button
              class="css-y6mgpv"
              type="submit"
            >
              <div
                class="css-tkevr6"
              >
                Register
              </div>
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="css-l5skqa"
    >
      <div
        class="css-18hh56p"
      >
        <div
          class="chakra-stack css-1o71ot0"
        >
          <a
            class="chakra-link css-epvm6"
            href="https://api.soul-network.com/docs"
            target="_blank"
          >
            API Docs
          </a>
          <a
            class="chakra-link css-epvm6"
            href="https://github.com/soul-project/login-portal"
            target="_blank"
          >
            Github
          </a>
          <a
            class="chakra-link css-epvm6"
            href="https://github.com/soul-project/soul"
            target="_blank"
          >
            Soul Network's Github
          </a>
          <a
            class="chakra-link css-epvm6"
            href="mailto:lws803@gmail.com"
            target="_blank"
          >
            Contact me
          </a>
        </div>
        <p
          class="chakra-text css-1m360k3"
        >
          Copyright © 
          2022
           Soul Network.
        </p>
      </div>
    </div>
  </div>
</div>
`;
